File: python/src/multi_agent_orchestrator/agents/bedrock_flows_agent.py
from typing import List, Dict, Any, Optional, Callable
from dataclasses import dataclass
import os
import json
import boto3
from multi_agent_orchestrator.utils import (Logger, conversation_to_dict)
from multi_agent_orchestrator.agents import (Agent, AgentOptions)
from multi_agent_orchestrator.types import (ConversationMessage, ParticipantRole)

# BedrockFlowsAgentOptions Dataclass
@dataclass
class BedrockFlowsAgentOptions(AgentOptions):
    flowIdentifier: str = None
    flowAliasIdentifier: str = None
    bedrock_agent_client: Optional[Any] = None
    enableTrace: Optional[bool] = False
    flow_input_encoder: Optional[Callable] = None
    flow_output_decoder: Optional[Callable] = None


# BedrockFlowsAgent Class
class BedrockFlowsAgent(Agent):

    def __init__(self, options: BedrockFlowsAgentOptions):
        super().__init__(options)

        # Initialize bedrock agent client
        if options.bedrock_agent_client:
            self.bedrock_agent_client = options.bedrock_agent_client
        else:
            if options.region:
                self.bedrock_agent_client = boto3.client(
                    'bedrock-agent-runtime',
                    region_name=options.region or os.environ.get('AWS_REGION')
                )
            else:
                self.bedrock_agent_client = boto3.client('bedrock-agent-runtime')

        self.enableTrace = options.enableTrace
        self.flowAliasIdentifier = options.flowAliasIdentifier
        self.flowIdentifier = options.flowIdentifier

        if options.flow_input_encoder is None:
            self.flow_input_encoder = self.__default_flow_input_encoder
        else:
            self.flow_input_encoder = options.flow_input_encoder

        if options.flow_output_decoder is None:
            self.flow_output_decoder = self.__default_flow_output_decoder
        else:
            self.flow_output_decoder = options.flow_output_decoder

    def __default_flow_input_encoder(self,
        input_text: str,
        **kwargs
    ) -> Any:
        """Encode Flow Input payload as a string."""
        return [
                    {
                        'content': {
                            'document': input_text
                        },
                        'nodeName': 'FlowInputNode',
                        'nodeOutputName': 'document'
                    }
                ]

    def __default_flow_output_decoder(self, response: Any, **kwargs) -> ConversationMessage:
        """Decode Flow output as a string and create ConversationMessage."""
        decoded_response = response
        return ConversationMessage(
            role=ParticipantRole.ASSISTANT.value,
            content=[{'text': str(decoded_response)}]
        )

    async def process_request(
        self,
        input_text: str,
        user_id: str,
        session_id: str,
        chat_history: List[ConversationMessage],
        additional_params: Optional[Dict[str, str]] = None
    ) -> ConversationMessage:
        try:
            response = self.bedrock_agent_client.invoke_flow(
                flowIdentifier=self.flowIdentifier,
                flowAliasIdentifier=self.flowAliasIdentifier,
                inputs=[
                    {
                        'content': {
                            'document': self.flow_input_encoder(self, input_text, chat_history=chat_history, user_id=user_id, session_id=session_id, additional_params=additional_params)
                        },
                        'nodeName': 'FlowInputNode',
                        'nodeOutputName': 'document'
                    }
                ],
                enableTrace=self.enableTrace
            )

            if 'responseStream' not in response:
                raise ValueError("No output received from Bedrock model")

            eventstream = response.get('responseStream')
            final_response = None
            for event in eventstream:
                Logger.info(event) if self.enableTrace else None
                if 'flowOutputEvent' in event:
                    final_response = event['flowOutputEvent']['content']['document']

            bedrock_response = self.flow_output_decoder(self, final_response)

            return bedrock_response

        except Exception as error:
            Logger.error(f"Error processing request with Bedrock: {str(error)}")
            raise error

File: typescript/src/agents/bedrockFlowsAgent.ts
import { BedrockAgentRuntimeClient, InvokeFlowCommand } from "@aws-sdk/client-bedrock-agent-runtime";
import { Agent, AgentOptions } from "./agent";
import {
  ConversationMessage,
  ParticipantRole
} from "../types";

  export interface BedrockFlowsAgentOptions extends AgentOptions {
    region?: string;
    flowIdentifier: string;
    flowAliasIdentifier: string;
    bedrockAgentClient?: BedrockAgentRuntimeClient;
    enableTrace?: boolean;
    flowInputEncoder?: (agent: Agent, inputText: string, ...kwargs: any) => any;
    flowOutputDecoder?: (agent: Agent, response: any) => any;
  }

  export class BedrockFlowsAgent extends Agent {

    /** Protected class members */
    protected flowIdentifier: string;
    protected flowAliasIdentifier: string;
    protected bedrockAgentClient: BedrockAgentRuntimeClient;
    protected enableTrace: boolean;
    private flowInputEncoder: (agent: Agent, inputText: string, params: any) => any; // Accepts any additional properties
    private flowOutputDecoder: (agent: Agent, response: any) => ConversationMessage;

    constructor(options: BedrockFlowsAgentOptions) {
      super(options);

      this.bedrockAgentClient = options.bedrockAgentClient ?? (
        options.region
          ? new BedrockAgentRuntimeClient({ region: options.region })
          : new BedrockAgentRuntimeClient()
      );

      this.flowIdentifier = options.flowIdentifier;
      this.flowAliasIdentifier = options.flowAliasIdentifier;
      this.enableTrace = options.enableTrace ?? false;
      if (options.flowInputEncoder) {
        this.flowInputEncoder = options.flowInputEncoder;
      } else {
        this.flowInputEncoder = this.defaultFlowInputEncoder;
      }

      if (options.flowOutputDecoder){
        this.flowOutputDecoder = options.flowOutputDecoder;
      } else {
        this.flowOutputDecoder = this.defaultFlowOutputDecoder;
      }
    }

    defaultFlowInputEncoder(agent: Agent, inputText: string, ..._kwargs: any): any {
      return  inputText
    }

    defaultFlowOutputDecoder(agent: Agent, response: any): ConversationMessage {
      return {
        role: ParticipantRole.ASSISTANT,
        content: [{ text: response }],
      };
    }

    async processRequest(
      inputText: string,
      userId: string,
      sessionId: string,
      chatHistory: ConversationMessage[],
      additionalParams?: Record<string, string>
    ): Promise<ConversationMessage> {

      try {
        // Prepare the command to invoke Bedrock Flows
        const invokeFlowCommand = new InvokeFlowCommand({
          flowIdentifier:this.flowIdentifier,
          flowAliasIdentifier:this.flowAliasIdentifier,
          inputs: [
            {
                'content': {
                    'document': this.flowInputEncoder(this, inputText, {userId:userId, sessionId:sessionId, chatHistory:chatHistory, ...additionalParams})
                },
                'nodeName': 'FlowInputNode',
                'nodeOutputName': 'document'
            }
          ],
          enableTrace: this.enableTrace
        });

        let completion;
        // Call Bedrock's Invoke Flow API
        const response = await this.bedrockAgentClient.send(invokeFlowCommand);

        if (!response.responseStream) {
          throw new Error("No output received from Bedrock Llows");
        }

        // Aggregate chunks of response data
        for await (const flowEvent of response.responseStream) {
          if (flowEvent.flowOutputEvent) {
            completion = flowEvent.flowOutputEvent.content.document;
          } else if (this.enableTrace) {
            // Log chunk event details if tracing is enabled
            this.logger ? this.logger.info("Flow Event Details:", JSON.stringify(flowEvent, null, 2)) : undefined;
          }
        }

        // Return the completed response as a Message object
        return this.flowOutputDecoder(this, completion);

      } catch (error: unknown) {  // Explicitly type error as unknown
      // Handle error with proper type checking
      const errorMessage = error instanceof Error
        ? error.message
        : 'Unknown error occurred';

      this.logger ? this.logger.error("Error processing request with Bedrock:", errorMessage) : undefined;
      throw new Error(`Error processing request with Bedrock: ${errorMessage}`);
    }
  }
}